### PR TITLE
feat: support storing app parameters in .argocd-app.yaml file

### PR DESCRIPTION
**Pull request by alexmt**
Closes https://github.com/argoproj/argo-cd/issues/4083

### CODE REVIEW

File: reposerver/repository/repository.go [hunk 1]
```diff
@@ -447,8 +448,49 @@ func newEnv(q *apiclient.ManifestRequest, revision string) *v1alpha1.Env {
...
+	}
+	*source = merged
+	return nil
+}
+
 // GetAppSourceType returns explicit application source type or examines a directory and determines its application source type
 func GetAppSourceType(source *v1alpha1.ApplicationSource, path string) (v1alpha1.ApplicationSourceType, error) {
+	err := mergeSourceParameters(source, path)
+	if err != nil {
+		return "", err
```
Comment (jannfis) [inline reposerver/repository/repository.go:491]: I think we should provide some more details when `mergeSourceParameters()` fails and let the user know that the reason of failure was actually a failed merge of `.argocd-app.yaml`, i.e.

`return "", fmt.Errorf("error while merging .argocd-app.yaml: %v", err)`

or something along these lines.

**Comment by jannfis** [comment]
OK, I've given it a little more thought. 

There are basically two general scenarios: The `Application` resource is part of another application and therefore stored in Git, or it is an ephemeral Application created by the CLI, UI or API and hance has no reference in Git. I think Argo CD Image Updater needs to support both, and the latter case already is supported.

For the former case, we need to find out which Argo CD application (`parent`) actually spawns the Application (`child`) resource the patch must be applied to, and I think we can just use the `app.kubernetes.io/instance` label on the `parent` resource for that. So we actually need to push into the repo of `parent`, so that it will generate a patched `child` Application resource.

Does that make sense?

**Comment by alexmt** [comment]
Good point @jannfis!
I did not think of the case when the application manifest itself is stored in Git.

In theory, it should be possible to modify YAML file that represents the application. In practice that looks very complicated since we might have a lot of scenarios: parent application might consist of YAML files, might use Helm/Ksonnet/Jsonnet, might use .argocd-app.yaml, etc. Instead of automatically detecting the required scenario I would let the user manually specify `writeback strategy`. E.g.

```yaml
writeback:
  # store parameters in crd
  crd: {}
  # update parameters in app manifest file
  appManifestFile:
    path: ./apps/guestbook-app.yaml
  # parameters file
  parametersFile: {}
```

what do you think?

**Comment by alexmt** [comment]
Renamed `.argocd-app.yaml` to `.argocd-source.yaml` . PTAL. @jessesuen , @jannfis

**Review by jannfis — APPROVED** [review]
LGTM, let's see what we can do out of it!

**Pull request merged**